BJS2-40457 Recommendation feature

drewbal opened the pull request.

drewbal committed:
BJS2-42316 Add new Recommendation

RecommendationController, recommendation service, validation, mappers. Covered with tests

lexsandr666 committed:
BJS2-40455 : Создание и настройка скиллов

drewbal committed:
Merge branch 'feature-BJS2-42316-add-new-recommendation' into feature-BJS2-40457

# Conflicts:
#	src/main/java/school/faang/user_service/service/SkillService.java

drewbal committed:
BJS2-40457 Add new Recommendation: review fixes

Skill validation method moved from SkillService to SkillValidation package.
UserRepository dependence removed from RecommendationService; UserService class with findUserById method created.
Tests applied.

drewbal committed:
BJS2-40457 Feat: Update Recommendation

Added methods in Controller and Service classes allowing the user to update existing recommendation.
Existing code optimized by refactoring certain logic in separate methods.

drewbal committed:
BJS2-40457 Feat: Delete Recommendation

Added methods in Controller and Service classes allowing the user to delete existing recommendation.
Existing code optimized by separating validation logic on Controller and Service patterns.

drewbal committed:
BJS2-40457 Fix: Add Recommendation review fixes

Change of logics and large refactoring of Add recommendation feature

drewbal committed:
Merge branch 'Dragon-master-stream7' into feature-BJS2-40457

# Conflicts:
#	src/main/java/school/faang/user_service/service/UserService.java
#	src/test/java/school/faang/user_service/service/UserServiceTest.java

drewbal committed:
BJS2-40457 Feat: Get All Recommendations for User

New method returns the list of all recommendations for user by his ID.

drewbal committed:
BJS2-40457 Feat: Get All Given Recommendations

New method returns the list of all recommendations given by author by his ID.

mighty-mhsl commented on src/main/java/school/faang/user_service/service/SkillOfferService.java:
```diff
@@ -0,0 +1,36 @@
+        return skillOfferRepository.findAllByUserId(userId);
+    }
+
+    public void saveSkillOffers(Recommendation recommendation) {
```
Это должно быть внутри RecommendationService, т.к. тут работа с рекомендацией гораздо больше, чем с оффером

mighty-mhsl replied:
Только метод сохранения сущности будет в этом сервисе

drewbal replied:
Ага. А еще я сделаю дополнительный метод, который будет сохранять скиллы при обновлении рекомендации из дто без вызова сущности.

mighty-mhsl commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,121 @@
+                recommendationDto.getContent()
+        );
+        Recommendation recommendation = getRecommendationById(recommendationDto.getId());
+        skillOfferService.deleteAllByRecommendationId(recommendation.getId());
```
Не понимаю, зачем это делать. Мы удалили все рекомендации, а потом их сразу сохраняем, при том, что только что эту сущность достали из БД и вообще никак не меняли офферы скиллов

drewbal replied:
Так, погоди. Для удаления старых рекоммендаций, которые были до обновления нам нужен id. Его можно взять как из дто, так и из сущности из БД. Сущность нам нужна, потому что метод saveSkillOffers работает с ней, т.к. в базу SkillOffer сохраняется запись с id рекоммендации. Для оптимизации можно создать еще один метод для сохранения скиллов updateSkillOffers, который будет работать с дто, и нам не понадобится даже вызывать сущность из бд.

drewbal replied:
Нет, все же понадобится. Метод по добавлению гаранта требует рекоммендацию, так как на одном из этапов надо добавить связку объектов скилла и рекоммендациию

mighty-mhsl commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,121 @@
+        skillOfferService.saveSkillOffers(recommendation);
+        skillService.addGuarantee(recommendationDto);
+
+        return recommendationMapper.toDto(getRecommendationById(recommendationDto.getId()));
```
А зачем нам ее опять из БД вынимать доп. запросом, если она у нас уже в переменной recommendation тут лежит?

drewbal replied:
мы возвращаем дто, данные в котором не меняются

mighty-mhsl commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,121 @@
+    public boolean delete(long id) {
+        recommendationServiceValidator.validateRecommendationExistsById(id);
+        recommendationRepository.deleteById(id);
+        return !recommendationRepository.existsById(id);
```
Зачем нам доп. запрос этот в БД, когда мы выше ТОЧНО удалили эту запись?

drewbal replied:
А вот тут наверное можно удалить и не возвращать никакого значения.

mighty-mhsl commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    public boolean delete(long id) {
+        recommendationServiceValidator.validateRecommendationExistsById(id);
```
А зачем? Если я пытаюсь даже удалить кривые данные (плюс я уже в контроллере все отвалидировал), то зачем мне еще раз это валидировать? Если данных нет таких в БД, то мы их все равно удаляем - их и не было

drewbal replied:
Да, наверное не надо умничать.

mighty-mhsl commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,121 @@
+        List<Recommendation> allRecommendations = new ArrayList<>();
+        Page<Recommendation> page;
+
+        do {
```
Pageable.unpaged()

Если это передать параметром в метод репозитория, то он сразу все вернет, и не придется кучу раз в БД ходить

drewbal replied:
Блин, читерство!!! я долго читал про этот чертов Pageable, но такого метода не увидел, ээх :)

mighty-mhsl commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,121 @@
+        List<Recommendation> allRecommendations = new ArrayList<>();
+        Page<Recommendation> page;
+
+        do {
```
То же самое

drewbal replied:
Йес

mighty-mhsl commented on src/main/java/school/faang/user_service/validation/recommendation/RecommendationControllerValidator.java:
```diff
@@ -0,0 +1,50 @@
+import school.faang.user_service.exception.DataValidationException;
+
+@Component
+public class RecommendationControllerValidator {
```
Я бы делал один валидатор для всего, а не несколько разных
Они вообще должны быть не в курсе про архитектуру, которую мы используем. Типа я должен этот валидатор быть способен откуда угодно дернуть, а не только из контроллера

drewbal replied:
рефакторинг - судьба моя :)

drewbal committed:
Merge branch 'Dragon-master-stream7' into feature-BJS2-40457

# Conflicts:
#	src/main/java/school/faang/user_service/service/UserService.java
#	src/test/java/school/faang/user_service/service/UserServiceTest.java

drewbal committed:
BJS2-40457 Fix: review fixes

Validators for recommendation controller and recommendation service composed in one.
saveSkillOffers method moved to RecommendationService.
Cycling dependency error fixed.
Some logic fixed in create, update and delete methods of RecommendationService.
Page operations optimized.

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+@Component
```
Если ты добавил аннотацию RestController то аннотация Component становится избыточной, ее можно смело удалять.

drewbal replied:
Руслан, ты попал в нашу путаницу :)) Дело в том, что Саша (Lexsandr), который разрабатывал фичу Skill и связанное с ней, в качестве эксперимента предложил мне сделать черрипик его коммита, но этот коммит включал весь код, который мне и подтянулся. Поэтому часть этих замечаний к его части, которую я не трогал :)) 
Он свою часть обещал вчера отправить на PR. Но в любом случае я дам ему ссылку на этот PR, чтобы он прочитал и уже фиксил. А свои комменты ниже я прокомментирую

RuslanZinovyev replied:
Андрей, у меня для тебя плохие новости, забудь о Саше, раз ты сделал черри пик его коммита, то теперь это твоя проблема и ты должен это исправить =) Если хочешь сделать что-то хорошо, делай сам ;)

drewbal replied:
Эти исправлю, больше комментариев не было :)) Успею смерджить

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/skill/SkillDtoValidation.java:
```diff
@@ -0,0 +1,27 @@
+public class SkillDtoValidation {
+    public final int MAX_TITLE_LENGTH = 64;
+
+    public void Validate(SkillDto skillDto) {
```
имена методов всегда начинаются с маленькой буквы и используют camel case

drewbal replied:
Саша (Lexsandr)

RuslanZinovyev replied:
Это твой PR, забудь про Сашу =) 
Исправлять его ошибки придется тебе.

drewbal replied:
Все бы ошибки были такие :)

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,41 @@
+@RequiredArgsConstructor
+public class RecommendationController {
+    private final RecommendationService recommendationService;
+    private final RecommendationValidator recommendationControllerValidator;
```
Если имя класса RecommendationValidator то почему имя переменной recomendationControllerValidator? Почему не recommendationValidator? Понятно что именование переменных это личное дело разработчика, но все же лучше руководствоваться одной логикой для всех переменных

drewbal replied:
Согласен. Я разделил валидацию на 2 класса, потом Влад объяснил, что лучше делать один, я их собрал в единый, но не везде по коду скорректировал имена переменных.
Поправлено.

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/recommendation/RecommendationValidator.java:
```diff
@@ -0,0 +1,96 @@
+
+    public void validateDto(RecommendationDto recommendationDto) {
+
+        if (recommendationDto.getAuthorId() == null || recommendationDto.getAuthorId() == 0) {
```
Мелочь, однако потенциальный ввод 0 не будет соответствовать предупреждению "Author id cannot be null" 
Число 0 не является null

drewbal replied:
Согласен. Из мелочей строится любой хороший результат :)
Скорректировал, отнес проверку на 0 и <0 в один блок

drewbal replied:
И еще раз переделал по комменту ниже. Теперь обе проверки включены в один блок благодаря исправленному сообщению исключения. Спасибо!

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/recommendation/RecommendationValidator.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    public void validateId(Long id) {
```
Нет смысла проверять id == 0 и id <= 0 дважды, можно это объединить

`public void validateId(Long id) {
    if (id == null || id <= 0) {
        throw new DataValidationException("Id must be a positive number");
    }
}`

drewbal replied:
Действительно. Так код становится намного компактнее. Сделано

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/recommendation/RecommendationValidator.java:
```diff
@@ -0,0 +1,96 @@
+
+        if (lastRecommendation.isPresent()) {
+            LocalDate lastRecommendationDate = lastRecommendation.get().getCreatedAt().toLocalDate();
+            if (lastRecommendationDate != null &&
```
Не проблема, но данную проверку на null можно опустить ибо поле createdAt не будет null в таблице Recommendation, Hibernate позаботится о том чтобы timestamp всегда был на месте. 

@CreationTimestamp
    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "created_at")
    private LocalDateTime createdAt;

drewbal replied:
Логично! Супер :)

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,111 @@
+
+import java.util.List;
+
+@Component
```
Аннотация Service  семантически здесь более уместна чем Component

drewbal replied:
Хорошо! Просто еще не добрался до лекции, где поясняется эта разница

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillOfferService.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.List;
+
+@Component
```
Для всех сервисов лучше использовать аннотацию Service

drewbal replied:
Готово

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private List<Recommendation> getAllRecommendationsByAuthorId(long authorId) {
+        Page<Recommendation> recommendations = recommendationRepository.findAllByReceiverId(authorId, Pageable.unpaged());
```
Подозреваю что ты планировал сделать вызов findAllByAuthorId() вместо findAllByReceiverId

drewbal replied:
Yep ;)

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,111 @@
+    private final SkillService skillService;
+
+    public RecommendationDto create(RecommendationDto recommendationDto) {
+        recommendationValidator.validateSkillAndTimeRequirementsForGuarantee(recommendationDto);
```
Ты делаешь валидацию дважды, один раз на уровне контроллера другой раз тут в сервисе, если ты валидируешь DTO тут, то я рекомендую удалить валидацию с контроллера. Сделай все в одном месте, сервис более подходящий класс для этого.

drewbal replied:
Я понимаю о чем ты. Но нужно немножко разъяснения. На контроллере согласно заданию проходит первичная валидация входящих данных (на null и невозможные значения типа <=0), а в сервисе мы валидируем данные согласно бизнес-логике метода. 
Правильно я понимаю, что всю валидацию (и первичную, и бизнес-логики) можно переместить в сервис?

RuslanZinovyev replied:
Нет, ты все правильно понимаешь но я видел что в контроллере ты используешь кастомный валидатор, а не средства фреймворка поэтому и оставил этот комментарий. На данном этапе я понимаю что другого выхода у тебя нет, так как мы создаем контроллер, но фактически создаем обыкновенный класс без специальных аннотаций. Учитывая что ты уже продвинутый я ожидаю что ты будешь создавать полноценные рест контроллеры вместо компонентов заглушек.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    public RecommendationDto update(RecommendationDto recommendationDto) {
+        recommendationValidator.validateSkillAndTimeRequirementsForGuarantee(recommendationDto);
```
Тоже самое, если делаешь валидацию тут, то удали ее с контроллера.

drewbal replied:
Сделал. А с другими методами как поступить? тоже валидацию с контроллера перенести в них?

RuslanZinovyev replied:
Попробую объяснить разницу между валидацией контроллера и сервиса. 
Контроллер может валидировать только то, что входящие данные соблюдают определенный формат, не содержат пустых полей, правильно указывают идентификатор или почтовый адрес, если речь идет о теле запроса. Настраивается это декларативно внутренними средствами самого фреймворка с помощью аннотаций Valid и Validated а также правил валидации для RequestBody и PathVariable с помощью (@NotNull, @Size, @Pattern, etc.)

Валидация на уровне сервиса валидирует бизнес логику твоего веб приложения и тут как раз ты уже пишешь свой класс валидации с нужными методами, который проверят существует ли юзер в базе данных и тд. Вот такой валидатор не должен быть на уровне контроллера, это задача сервиса. Контроллер валидирует просто входные данные, что нет никаких ошибок, защита от дурака. Надеюсь смысл понятен, если нет напиши обсудим.

drewbal replied:
Супер понятно! Огромное спасибо! Толкнул меня в бездну валидации Spring, но рад, что познакомился с этим :)

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/RecommendationService.java:
```diff
@@ -0,0 +1,111 @@
+    private final SkillOfferService skillOfferService;
+    private final SkillService skillService;
+
+    public RecommendationDto create(RecommendationDto recommendationDto) {
```
Учитывая что методы create() и update() делают несколько манипуляций с базой данных, рассмотри использование аннотации Transactional чтобы в случае какой-либо ошибки операция откатилась и база данных была в консистентном состоянии. Если этот момент не ясен я тебе дополнительно объясню.

drewbal replied:
Ооо, классно! У меня как раз возникали мысли, как быть, если вдруг что-то пойдет не так, и одна из операция метода с БД не выполнится. Теперь прочитал и буду использовать :)

RuslanZinovyev replied:
Не забывай закрывать комментарии если уже исправил что-то. Так легко понять что было сделано, а что нет. Нажимай Resolve conversation.

RuslanZinovyev commented on src/main/java/school/faang/user_service/service/SkillService.java:
```diff
@@ -0,0 +1,71 @@
+                .toList();
+    }
+
+    public void addGuarantee(Recommendation recommendation) {
```
Рассмотри использование аннотации Transactional чтобы использовать декларативный способ создания транзакции, в этом методе несколько манипуляций с базой данных

drewbal replied:
Добавил

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/recommendation/RecommendationServiceValidator.java:
```diff
@@ -0,0 +1,55 @@
+
+@Component
+@RequiredArgsConstructor
+public class RecommendationServiceValidator {
```
Нет смысла разделять валидторы для сервиса и контроллера, по хорошему должен быть один RecomendationValidator класс и использовать его лучше на уровне сервиса.

drewbal replied:
Уж пофиксил, но забыл удалить старый :)

drewbal committed:
BJS2-40457 Fix: review fixes

Validation moved from controller to service;
@Transactional annotations used for some methods;
Data format checks in RecommendationValidator optimized;

denis-levkhov commented on src/main/java/school/faang/user_service/validation/recommendation/RecommendationValidator.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public void validateId(Long id) {
+        if (id == null || id <=0) {
```
Пропустил пробел после `<=`

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/RecommendationController.java:
```diff
@@ -0,0 +1,39 @@
+@RequiredArgsConstructor
+public class RecommendationController {
+    private final RecommendationService recommendationService;
+    private final RecommendationValidator recommendationValidator;
```
Нам не нужен самописный валидатор на уровне контроллера, я подробный комментарий оставил в ревью, на данном этапе тебе не очевидно это потому что несмотря на то что класс называется контроллер он не использует много фич контроллера, которые мы расмотрим позже, таких как @PathVariable и тд

drewbal replied:
Да, сэнсей :) Поизучал и решил попробовать. Долой ручную валидацию из контроллера! :) Нужно мнение, правильно ли?

drewbal committed:
Merge branch 'Dragon-master-stream7' into feature-BJS2-40457

drewbal committed:
BJS2-40457 Fix: review fixes

Controller validation switched from custom validation to RestController Spring annotations.
Subsequent changes in DTOs, RecommendationValidator and RecommendationService classes.

drewbal committed:
BJS2-40457 Fix: review fixes

Controller validation switched from custom validation to RestController Spring annotations.
Subsequent changes in DTOs, RecommendationValidator and RecommendationService classes.

drewbal committed:
BJS2-40457 Fix: review fixes

Controller validation switched from custom validation to RestController Spring annotations.
Subsequent changes in DTOs, RecommendationValidator and RecommendationService classes.

drewbal committed:
Merge remote-tracking branch 'origin/feature-BJS2-40457' into feature-BJS2-40457

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,46 @@
+@RequiredArgsConstructor
+public class SkillController {
+    private final SkillService skillService;
+    private final SkillDtoValidation skillDtoValidation;
```
Все хорошо но снова валидатор на контроллере глаз режет =) 
Ты входящие данные DTO можешь провалидировать аннотациями, (mix, max, etc) Validated на контроллере будет проверять это, ты сам можешь REST клиентом локально потестить
Также SkillDtoValidation не удачное название класса, просто SkillValidation достаточно.

drewbal replied:
А "чужой" код править даже весело :)

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/SkillController.java:
```diff
@@ -0,0 +1,46 @@
+        return skillService.getUserSkills(userId);
+    }
+
+    public SkillCandidateDto getOfferedSkills(SkillCandidateDto skillCandidateDto) {
```
как обращаться к этому ендпоинту? где URL?

drewbal replied:
Кроме этого переписал логику метода в контроллере и сервисе

RuslanZinovyev commented on src/main/java/school/faang/user_service/dto/skill/SkillDto.java:
```diff
@@ -0,0 +1,12 @@
+@Data
+@AllArgsConstructor
+public class SkillDto {
+    private Long id;
```
Повесь validation constraint тут и тогда тебе не нужно будет валидировать этот DTO своим самописным валидатором на уровне контроллера.

drewbal replied:
Легко!

RuslanZinovyev commented on src/main/java/school/faang/user_service/dto/skill/SkillCandidateDto.java:
```diff
@@ -0,0 +1,12 @@
+
+@Data
+@AllArgsConstructor
+public class SkillCandidateDto {
```
Вот смотри, такой же DTO как и другие, а тут нет validation constraint в отличие от других DTO, почему?

drewbal replied:
Теперь красиво :)

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/skill/SkillDtoFieldConstraintValidationException.java:
```diff
@@ -0,0 +1,14 @@
+
+public class SkillDtoFieldConstraintValidationException extends RuntimeException{
+
+    private static final Logger logger = LoggerFactory.getLogger(SkillDtoFieldConstraintValidationException.class);
```
Тут ошибки нет, но если ты уже пользуешь Lombok библиотеку в проекте можно создать логгер повесив простую аннотацю Slf4j на уровень класса и имя лоллера будет log. Можешь сам оценить магию.

drewbal replied:
полностью согласен

drewbal replied:
Я его позже вообще удалил, так как валидацию делаем через аннотации Spring

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/skill/SkillDtoNullObjectValidationException.java:
```diff
@@ -0,0 +1,14 @@
+
+public class SkillDtoNullObjectValidationException extends RuntimeException{
+
+    private static final Logger logger = LoggerFactory.getLogger(SkillDtoNullObjectValidationException.class);
```
Если логгер Lombok понравился то можешь его использовать и тут, но как я и сказал если оставишь как есть ошибкой не будет.

drewbal replied:
Заменил

drewbal replied:
Да тоже удалил. И @Slf4j тут не нужен, мы не логируем внутри исключения

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/skill/SkillDuplicateException.java:
```diff
@@ -0,0 +1,14 @@
+
+    public SkillDuplicateException(String message) {
+        super(message);
+        logger.error(message);
```
В этом не особо много смысла, все что ты тут делаешь это дублируешь тоже самое сообщение, которое и так будет выброшено в консоль исключением. В данном случае я рекомендую добавить логгер в сервис, который может потенциально выбросить данное исключение и залогировать стек трейс, чтобы ты по логу мог понять где конкретно возникла проблема. Поверь, со временем ты поймешь как это важно.

drewbal replied:
Я для пробы сделал логирование в сервисе с дальнейшим пробросом. По идее в сервису промежуточная обработка не очень нужна, надо где-то выше в глобальном ExceptionHandler, но его еще не писали (пожалуйста, давай не будем в этой задаче) :))))

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/skill/SkillDtoValidation.java:
```diff
@@ -0,0 +1,27 @@
+
+@Component
+@RequiredArgsConstructor
+public class SkillDtoValidation {
```
Просто SkillValidation и соедини логику двух валидаторов SkillDtoValidation and SkillValidation в один класс, кстати более уместное название будет SkillValidator - классы по своей сути существительные =)

drewbal replied:
SkillDtoValidation я удалил, так как валидацию контроллером делаем по-другому. Второе название изменил, рад, что существительные - правильный подход

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/skill/SkillValidation.java:
```diff
@@ -0,0 +1,23 @@
+
+    public void validateDuplicate(Skill skill) {
+        if(skillRepository.existsByTitle(skill.getTitle())) {
+            throw new SkillDuplicateException("Такой навык уже существует");
```
Настоятельно рекомендую писать все исключения на английском языке, за грамматику не переживай.

drewbal replied:
Absolutely

RuslanZinovyev commented on src/main/java/school/faang/user_service/validation/user/UserValidator.java:
```diff
@@ -0,0 +1,18 @@
+
+@Component
+@RequiredArgsConstructor
+public class UserValidator {
```
Вот тут название класса по красоте, заметь Андрей у тебя нет постоянства, один класс назвал Validation, другой Validator, нужно придерживаться одной стратегии в именованиее переменных и классов.

drewbal replied:
У меня постоянство есть ;)

RuslanZinovyev commented on src/test/java/school/faang/user_service/service/RecommendationServiceTest.java:
```diff
@@ -0,0 +1,231 @@
+        RecommendationDto result = recommendationService.update(dto);
+        result.setContent("Updated content");
+
+        assertNotNull(result);
```
рекомендую ассерты всегда указывать в самом низу, разработчики глазами их ищут всегда внизу

drewbal replied:
Скорректировал. Изменю подход. Почему-то я решил сначала ставить ассерты, а после них блок verify. Поменял местами

drewbal committed:
BJS2-40457 Fix: review fixes

Skill related part of the code fixed:
Controller REST annotations and validation fixed;
Exceptions fixed (unneccessary deleted due to new validation approach);
getOfferedSkills method applies new correct logic.
SkillDtoValidation class deleted (new validation approach)

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/skill/SkillDuplicateException.java:
```diff
@@ -0,0 +1,11 @@
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
```
тогда и отсюда надо логгер удалить раз ты в исключении ничего логировать не будешь

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/skill/SkillDtoNullObjectValidationException.java:
```diff
@@ -0,0 +1,14 @@
+
+    public SkillDtoNullObjectValidationException(String message) {
+        super(message);
+        logger.error(message);
```
логгер на месте как я вижу

RuslanZinovyev commented on src/main/java/school/faang/user_service/exception/skill/SkillDtoNullObjectValidationException.java:
```diff
@@ -0,0 +1,14 @@
+
+public class SkillDtoNullObjectValidationException extends RuntimeException{
+
+    private static final Logger logger = LoggerFactory.getLogger(SkillDtoNullObjectValidationException.class);
```
можно удалить

RuslanZinovyev replied:
кстати имя исключения достаточно пугающее, что конкретно оно делает и в каком случае вылетает?

drewbal replied:
Уже ни в каком. Оно было для ручной валидации, но я это исправил. Я думал, что удалил его. Теперь точчно удалил

drewbal committed:
BJS2-40457 Fix: review fixes

Exceptions: Exception for validation purposes was removed.
Other: minor code cleaning works

drewbal committed:
Merge branch 'Dragon-master-stream7' into feature-BJS2-40457

# Conflicts:
#	src/main/java/school/faang/user_service/exception/DataValidationException.java

drewbal committed:
BJS2-40457 Fix: review fixes

UserService: imports fixed
